---
title: "DEG_analysis"
author: "Adrian Salas"
date: "April 18, 2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


## The next steps are intended for computing and analyzing the differentially expressed genes using the **RaceID** built in functions.

Loading the environment where the data was stored:
```{r}
load("~/Documents/Bioinformatic_Analysis/R_workspaces_sc/Base_Markdowns.RData", verbose = T)
```

Load the requiered libraries.

```{r}
library(dplyr)
library(RaceID)
```

## Differential gene expression analysis between clusters
Create a list and store all the different outputs after computing DEGs for each cluster using clustdiffgenes() function.
```{r}
DEGsAll <- list()

for (i in 1:max(sc@cpart)) {
  DEGsAll[[i]] <- clustdiffgenes(sc, i, pvalue = 1)
}

```

Filter the results for each cluster based on the p value threshold (p < 0.05) and store all the different outputs as a list.
```{r}
DEGsAllFTD <- DEGsAll
for (i in 1:length(DEGsAll)) {
  DEGsAllFTD[[i]]$gene <- rownames(DEGsAll[[i]])
  DEGsAllFTD[[i]]$log2FoldChange <- log2(DEGsAll[[i]]$fc)
  DEGsAllFTD[[i]] <- filter(DEGsAllFTD[[i]], padj < 0.05)
  DEGsAllFTD[[i]] <- DEGsAllFTD[[i]][ , c(6,3,7,4,5)]
}

```

## Differential gene expression analysis per cluster between conditions
Create a list to store the information of the cells belonging to the different clusters divided by condition.
```{r}
cellsInClCond <- list()
for (i in (1:max(sc@cpart))){
  cellsin <- names(sc@cpart[sc@cpart == i])
  cellsCont <- cellsin[grep("^C", cellsin)]
  cellscKO <- cellsin[grep("^K|k", cellsin)]
  cellsInClCond[[i]] <- list(cellsCont, cellscKO)  
} 
#####  CHECK THIS OIECE OF CODE BECAUSE I SUPPOSE THE CELL NAMES ARE NOT THE RIGHT ONES !!!!
  
```

Compute the DEGs per cluster between conditions
```{r}
DEG_all_Cond <- list()
for (i in 1:max(sc@cpart)){  
  if (length(cellsInClCond[[i]][[1]]) != 0 & length(cellsInClCond[[i]][[2]]) != 0) {
  A <- cellsInClCond[[i]][[1]] # control cells
  B <- cellsInClCond[[i]][[2]] # cKO cells
  DEG_all_Cond[[i]] <- diffexpnb(getfdata(sc, n=c(A,B)), A=A, B=B)
  }
  else
    DEG_all_Cond[[i]] <- "No DEGs computed"
}
```

Filter the results for each cluster based on the p value threshold (p < 0.05) and store all the different outputs as a list.
```{r}
DEG1cond <- DEG_all_Cond[[1]][[3]][4:7] 
DEG1cond$gene <- rownames(DEG1cond) 
DEG1condFtd <- filter(DEG1cond, padj < 0.05)

DEG2cond <- DEG_all_Cond[[2]][[3]][4:7]
DEG2cond$gene <- rownames(DEG2cond) 
DEG2condFtd <- filter(DEG2cond, padj < 0.05)

DEG3cond <- DEG_all_Cond[[3]][[3]][4:7]
DEG3cond$gene <- rownames(DEG3cond) 
DEG3condFtd <- filter(DEG3cond, padj < 0.05)

DEG4cond <- DEG_all_Cond[[4]][[3]][4:7]
DEG4cond$gene <- rownames(DEG4cond) 
DEG4condFtd <- filter(DEG4cond, padj < 0.05)

DEG5cond <- DEG_all_Cond[[5]][[3]][4:7] 
DEG5cond$gene <- rownames(DEG5cond) 
DEG5condFtd <- filter(DEG5cond, padj < 0.05)

DEG6cond <- DEG_all_Cond[[6]][[3]][4:7] 
DEG6cond$gene <- rownames(DEG6cond) 
DEG6condFtd <- filter(DEG6cond, padj < 0.05)

DEG7cond <- DEG_all_Cond[[7]][[3]][4:7] 
DEG7cond$gene <- rownames(DEG7cond) 
DEG7condFtd <- filter(DEG7cond, padj < 0.05)

DEG8cond <- DEG_all_Cond[[8]][[3]][4:7]
DEG8cond$gene <- rownames(DEG8cond) 
DEG8condFtd <- filter(DEG8cond, padj < 0.05)

DEG9cond <- DEG_all_Cond[[9]][[3]][4:7] 
DEG9cond$gene <- rownames(DEG9cond) 
DEG9condFtd <- filter(DEG9cond, padj < 0.05)

DEG10cond <- DEG_all_Cond[[10]][[3]][4:7] 
DEG10cond$gene <- rownames(DEG10cond) 
DEG10condFtd <- filter(DEG10cond, padj < 0.05)

DEG11cond <- DEG_all_Cond[[11]][[3]][4:7] 
DEG11cond$gene <- rownames(DEG11cond) 
DEG11condFtd <- filter(DEG11cond, padj < 0.05)

DEG12cond <- DEG_all_Cond[[12]][[3]][4:7] 
DEG12cond$gene <- rownames(DEG12cond) 
DEG12condFtd <- filter(DEG12cond, padj < 0.05)

DEG13cond <- DEG_all_Cond[[13]][[3]][4:7] 
DEG13cond$gene <- rownames(DEG13cond) 
DEG13condFtd <- filter(DEG13cond, padj < 0.05)

DEG14cond <- DEG_all_Cond[[14]][[3]][4:7] 
DEG14cond$gene <- rownames(DEG14cond) 
DEG14condFtd <- filter(DEG14cond, padj < 0.05)

DEG15cond <- DEG_all_Cond[[15]][[3]][4:7] 
DEG15cond$gene <- rownames(DEG15cond) 
DEG15condFtd <- filter(DEG15cond, padj < 0.05)

DEG16cond <- DEG_all_Cond[[16]][[3]][4:7] 
DEG16cond$gene <- rownames(DEG16cond) 
DEG16condFtd <- filter(DEG16cond, padj < 0.05)

DEG17cond <- DEG_all_Cond[[17]][[3]][4:7] 
DEG17cond$gene <- rownames(DEG17cond) 
DEG17condFtd <- filter(DEG17cond, padj < 0.05)

DEG18cond <- DEG_all_Cond[[18]][[3]][4:7] 
DEG18cond$gene <- rownames(DEG18cond) 
DEG18condFtd <- filter(DEG18cond, padj < 0.05)

DEG20cond <- DEG_all_Cond[[20]][[3]][4:7]
DEG20cond$gene <- rownames(DEG20cond) 
DEG20condFtd <- filter(DEG20cond, padj < 0.05)

DEG21cond <- DEG_all_Cond[[21]][[3]][4:7] 
DEG21cond$gene <- rownames(DEG21cond) 
DEG21condFtd <- filter(DEG21cond, padj < 0.05)

DEG23cond <- DEG_all_Cond[[23]][[3]][4:7] 
DEG23cond$gene <- rownames(DEG23cond) 
DEG23condFtd <- filter(DEG23cond, padj < 0.05)

DEG24cond <- DEG_all_Cond[[24]][[3]][4:7] 
DEG24cond$gene <- rownames(DEG24cond) 
DEG24condFtd <- filter(DEG24cond, padj < 0.05)

DEG25cond <- DEG_all_Cond[[25]][[3]][4:7]
DEG25cond$gene <- rownames(DEG25cond) 
DEG25condFtd <- filter(DEG25cond, padj < 0.05)

DEG_all_CondFTD <- list(DEG1condFtd, DEG2condFtd,DEG3condFtd,DEG4condFtd,DEG5condFtd,DEG6condFtd,
           DEG7condFtd,DEG8condFtd,DEG9condFtd,DEG10condFtd,DEG11condFtd,DEG12condFtd,
           DEG13condFtd,DEG14condFtd,DEG15condFtd,DEG16condFtd,DEG17condFtd,DEG18condFtd,
           DEG20condFtd,DEG21condFtd,DEG23condFtd,DEG25condFtd)

```

Save the gene names for the DEGs from each comparison in a vector, and create a table with the frequencies
```{r}
genes <- c(DEG1condFtd$gene, DEG2condFtd$gene,DEG3condFtd$gene,DEG4condFtd$gene,DEG5condFtd$gene,DEG6condFtd$gene,
           DEG7condFtd$gene,DEG8condFtd$gene,DEG9condFtd$gene,DEG10condFtd$gene,DEG11condFtd$gene,DEG12condFtd$gene,
           DEG13condFtd$gene,DEG14condFtd$gene,DEG15condFtd$gene,DEG16condFtd$gene,DEG17condFtd$gene,DEG18condFtd$gene,
           DEG20condFtd$gene,DEG21condFtd$gene,DEG23condFtd$gene,DEG25condFtd$gene)

geneCounts <- as.data.frame(table(genes))
head(geneCounts)
```

Filter the genes obtained based on a "Black List" of genes reported as sex biased.
```{r}
BlackListDF <- read.csv("~/Documents/Bioinformatic_Analysis/Sex_biased_genes2014_datasets/blackListSexGenes.csv")
BlackList <- BlackListDF$Symbol
BlackList <- as.data.frame(BlackList[-c(32,46,91,92,93)])
colnames(BlackList) <- "gene"

geneCounts[(geneCounts$genes %in% BlackList$gene), ]

geneCountsBLFtd <- geneCounts[!(geneCounts$genes %in% BlackList$gene), ]

```


Plotting all the DEGs between conditions as a heatmap
```{r, echo=TRUE}
library(ComplexHeatmap)
library(circlize)

# Full join all the different dataframes but just the important columns for the heatmap
DEG_DF <- DEG_all_CondFTD[[1]][c(2,5)]
for (i in 2:length(DEG_all_CondFTD)) {
  DEG_DF <- full_join(DEG_DF, DEG_all_CondFTD[[i]][c(2,5)], by= "gene", keep=T)
} 

# Removing the NAs from the dataframe
for (i in 1:nrow(DEG_DF)) {
  for (j in 1:ncol(DEG_DF)) {
    if (is.na(DEG_DF[i,j])) {
      DEG_DF[i,j] <- 0
    }
  }
}

# Simplify a bit more the data frma
# Set rownames
rownames(DEG_DF) <- DEG_DF$gene
# Remove gene column
DEG_DF <- DEG_DF[ , -2]
# Set the colnames
#colnames(DEG_DF) <- paste0("cl", c(1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,20,21,23,25))
celltypes <- c("Granule", "CA1pyr", "NSC1", "CA3pyr","IPC1","Subic1", "CajRet1", "NCS2", "IPC2","CajRet2", "NSC3", "NSC4cyc", "Subic2", "INs", "Pericytes", "Unk1", "CortHem", "CajRet3", "Unk2", "NSC5", "CajRet4")
colnames(DEG_DF) <- celltypes

# Removing "Xist" gene
DEG_DF <- DEG_DF[!rownames(DEG_DF) == "Xist", ]
# Changing the dataframe to a matrix for the input
DEG_DF <- as.matrix(DEG_DF)

# Check and filter out columns that don't have DEGs between conditions
colSums(DEG_DF)
#DEG_DF <- DEG_DF[ , -15]

# Generating a simple heatmap
set.seed(1500)
Heatmap(DEG_DF, show_row_names = F, row_title = "genes", column_title_side = "top", show_heatmap_legend = T, name = "log2FC", col = colorRamp2(c(-7, 0, 7), c("blue", "white", "red")), column_split = 2, column_title = "%s", border = T)

# Heatmap(DEG_DF, show_row_names = F, row_title = "genes", column_title = "%s", column_title_side = "top", show_heatmap_legend = T, name = "log2FC", col = colorRamp2(c(-7, 0, 7), c("blue", "white", "red")), column_km = 2, cluster_rows = T, border = T)
```


Plotting "main" DEGs between conditions as a heatmap.
```{r, echo=TRUE}

# Filtering for the DEGs that appear significant in at least 5 different clusters.
geneCountsFtd5 <- filter(geneCountsBLFtd, Freq >= 5) 

Heatmap(DEG_DF[rownames(DEG_DF) %in% geneCountsFtd5$genes, -c(14,17,18)], show_row_names = T, row_title = "genes", column_title_side = "top", show_heatmap_legend = T, name = "log2FC", col = colorRamp2(c(-4, 0, 4), c("blue", "white", "red")), column_split = 4, column_title = "%s", border = T)

# Heatmap(DEG_DF[rownames(DEG_DF) %in% geneCountsFtd5$genes, -c(14,17,18)], show_row_names = T, row_title = "genes", column_title_side = "top", show_heatmap_legend = T, name = "log2FC", col = colorRamp2(c(-4, 0, 4), c("blue", "white", "red")), column_km = 4, column_km_repeats = 150, column_title = "%s", border = T)
```

Save the environment for the following analysis.
```{r}
save.image("~/Documents/Bioinformatic_Analysis/R_workspaces_sc/Base_Markdowns.RData")
```
