---
title: "Dimensionality reduction and clustering with RaceID algorithm"
author: "Adrian Salas"
date: "4/15/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

The following steps are used for processing the scRNAseq output matrix, including filtering, dimensions reduction and clustering with the **RaceID** algorithm version 0.1.3.


Loading the environment where the data was stored:
```{r}
load("~/Documents/Bioinformatic_Analysis/R_workspaces_sc/Base_Markdowns.RData", verbose = T)
```

Loading the required packages:
```{r}
library(RaceID)
```
Remove ERCC and mitochondrial genes

```{r}
prdata           <- as.data.frame(merged_data_tib)
rownames(prdata) <- prdata$GENEID
prdata$GENEID    <- NULL
prdata           <- prdata[grep("^(ERCC|mt|Gm|Rik)",row.names(prdata),invert=TRUE),]
```

Remove cells based on transcript counts (less than or equal to 500 total transcripts)
```{r}
cs <- colSums(prdata)
prdata <- prdata[,cs > 500]

```

Initialize the "sc" object
```{r}
sc <- SCseq(prdata)
```


Filter the cells with at least 'mintotal' number of  total transcripts
```{r}
sc <- filterdata(sc, 
                 mintotal     =  3000, 
                 minexpr      =  5, 
                 minnumber    =  1,
                 LBatch       =  NULL, 
                 knn          =  10, 
                 CGenes       =  c("Mki67", "Pcna"), 
                 FGenes       =  c("Kcnq1ot1"),
                 ccor         =  0.4,
                 bmode        =  "RaceID")
```

Compute cell distances
```{r}
sc <- compdist(sc, 
               metric        =  "pearson", 
               FSelect       =  T, 
               knn           =  NULL)
```

Perform cells clustering. The "cln" parameter value has to be defined in a data set dependent manner. See plots below for an indication on how to define the parameter.
```{r}
sc <- clustexp(sc, 
               sat          =  TRUE, 
               cln          =  16, 
               clustnr      =  30,
               bootnr       =  50, 
               rseed        =  17000, 
               FUNcluster   =  "kmedoids")
```

Perform dimensionality reduction using t-SNE representation
```{r}
sc <- comptsne(sc, initial_cmd = TRUE, perplexity = 30, rseed = 15555)
```

k-nearest neighbour graph layout utilizing the Fruchterman-Rheingold algorithm
```{r}
sc <- compfr(sc,knn=10)

```

Run outlier identification
```{r}
sc <- findoutliers(sc, 
                   probthr      =  1e-5, 
                   outminc      =  3, 
                   outlg        =  3,
                   outdistquant = 0.95)
```


## Plots

Plot t-SNE representation of the data
```{r, echo=TRUE}
plotmap(sc)
```

Plot of change in log within cluster dispersion at different values of k. The results of this plot can help to define the "cln" parameter for the clustexp() function.
```{r, echo=TRUE}
plotsaturation(sc)
```

Plot Jaccard's similarity for checking clusters stability.
```{r, echo=TRUE}
plotjaccard(sc)
```

Save the environment for the following analysis. 
```{r}
save.image("~/Documents/Bioinformatic_Analysis/R_workspaces_sc/Base_Markdowns.RData")
```
